File: discord-bot-fleet/src/tools/fetch_url.ts
// fetch_url — fetches a URL and returns sanitized text content.
// Strips HTML tags, limits to 5000 chars, no JavaScript execution.

import type { LLMTool } from '../llm.js';

export const fetchUrlTool: LLMTool = {
  type: 'function',
  function: {
    name: 'fetch_url',
    description: 'Fetch a URL and return the text content (HTML stripped). Useful for reading articles, documentation, or API responses. Limited to 5000 chars.',
    parameters: {
      type: 'object',
      properties: {
        url: { type: 'string', description: 'The URL to fetch' },
        max_chars: { type: 'number', description: 'Max chars to return (default 5000, max 10000)' },
      },
      required: ['url'],
    },
  },
};

export async function fetchUrlHandler(args: { url: string; max_chars?: number }): Promise<string> {
  const maxChars = Math.min(args.max_chars || 5000, 10000);
  const controller = new AbortController();
  const timeout = setTimeout(() => controller.abort(), 10000);
  try {
    const res = await fetch(args.url, {
      signal: controller.signal,
      redirect: 'follow',
      headers: { 'User-Agent': 'Mozilla/5.0 (compatible; DiscordBot/1.0)' },
    });
    if (!res.ok) return JSON.stringify({ error: `HTTP ${res.status} ${res.statusText}`, url: args.url });
    const ct = res.headers.get('content-type') || '';
    const body = await res.text();
    let text: string;
    if (ct.includes('application/json')) {
      text = body;
    } else if (ct.includes('text/html')) {
      // Strip tags, scripts, styles, decode basic entities
      text = body
        .replace(/<script[\s\S]*?<\/script>/gi, '')
        .replace(/<style[\s\S]*?<\/style>/gi, '')
        .replace(/<[^>]+>/g, ' ')
        .replace(/&nbsp;/g, ' ')
        .replace(/&amp;/g, '&')
        .replace(/&lt;/g, '<')
        .replace(/&gt;/g, '>')
        .replace(/&quot;/g, '"')
        .replace(/&#39;/g, "'")
        .replace(/\s+/g, ' ')
        .trim();
    } else {
      text = body;
    }
    return text.slice(0, maxChars);
  } catch (e: unknown) {
    const err = e as Error;
    return JSON.stringify({ error: err.message || String(e), url: args.url });
  } finally {
    clearTimeout(timeout);
  }
}

File: bot-python/tools/fetch_url.py
"""fetch_url — fetches a URL and returns sanitized text. SSRF-protected."""

import json
import socket
import ipaddress
import re
from urllib.parse import urlparse
import aiohttp

FETCH_URL_TOOL = {
    'type': 'function',
    'function': {
        'name': 'fetch_url',
        'description': 'Fetch a URL and return the text content (HTML stripped). Useful for reading articles, documentation, or API responses. Limited to 5000 chars. Cannot access private/internal IP ranges.',
        'parameters': {
            'type': 'object',
            'properties': {
                'url': {'type': 'string', 'description': 'The URL to fetch'},
                'max_chars': {'type': 'number', 'description': 'Max chars to return (default 5000, max 10000)'},
            },
            'required': ['url'],
        },
    },
}


PRIVATE_HOSTS = {
    'localhost', 'metadata.google.internal', 'metadata.azure.com',
    '169.254.169.254',
}


def is_private_host(hostname: str) -> bool:
    host = hostname.strip('[]').lower()
    if host in PRIVATE_HOSTS:
        return True
    if host.endswith('.internal') or host.endswith('.local'):
        return True
    try:
        ip = ipaddress.ip_address(host)
        return ip.is_private or ip.is_loopback or ip.is_link_local or ip.is_reserved or ip.is_multicast
    except ValueError:
        pass
    # DNS lookup to catch rebinding attacks
    try:
        results = socket.getaddrinfo(host, None)
        for r in results:
            ip_str = r[4][0]
            try:
                ip_obj = ipaddress.ip_address(ip_str)
                if ip_obj.is_private or ip_obj.is_loopback or ip_obj.is_link_local or ip_obj.is_reserved:
                    return True
            except ValueError:
                continue
    except Exception:
        pass
    return False


async def fetch_url_handler(args, ctx):
    url = args.get('url', '').strip()
    if not url:
        return json.dumps({'error': 'url required'})
    max_chars = min(int(args.get('max_chars', 5000)), 10000)
    try:
        parsed = urlparse(url)
    except Exception:
        return json.dumps({'error': f'invalid URL: {url}'})
    if parsed.scheme not in ('http', 'https'):
        return json.dumps({'error': f'only http/https allowed, got: {parsed.scheme}'})
    if not parsed.hostname:
        return json.dumps({'error': 'no hostname in URL'})
    if is_private_host(parsed.hostname):
        return json.dumps({'error': f'blocked: {parsed.hostname} is a private/internal address'})

    try:
        async with aiohttp.ClientSession(
            timeout=aiohttp.ClientTimeout(total=10),
            headers={'User-Agent': 'Mozilla/5.0 (compatible; DiscordBot/1.0)'},
        ) as session:
            async with session.get(url, allow_redirects=True) as resp:
                if resp.status != 200:
                    return json.dumps({'error': f'HTTP {resp.status} {resp.reason}', 'url': url})
                ct = resp.headers.get('content-type', '')
                body = await resp.text()
        if 'application/json' in ct:
            text = body
        elif 'text/html' in ct:
            text = re.sub(r'<script[\s\S]*?</script>', '', body, flags=re.IGNORECASE)
            text = re.sub(r'<style[\s\S]*?</style>', '', text, flags=re.IGNORECASE)
            text = re.sub(r'<[^>]+>', ' ', text)
            text = (text.replace('&nbsp;', ' ').replace('&amp;', '&')
                        .replace('&lt;', '<').replace('&gt;', '>')
                        .replace('&quot;', '"').replace('&#39;', "'"))
            text = re.sub(r'\s+', ' ', text).strip()
        else:
            text = body
        return text[:max_chars]
    except Exception as e:
        return json.dumps({'error': str(e), 'url': url})
